### DISCUSSION TITLE
netbox DynamicModelMultipleChoiceField validation error

### DISCUSSION THREAD

**Discussion by aakozlov85** [netbox-community/netbox#13181] [Help Wanted!]
Hello community!
Right now i am developing netbox plugin and trying to acomplish task where i can choose a few devices in form and add them to database table.
To do that i am using Model Form and DynamicModelMultipleChoiceField.
It is something wrong in my code so form can't be validated in correct way.
Get an error “<ConfigContextModelQuerySet [<Device: Phone5>]>” value must be an integer.

When i use DynamicModelChoiceField i have no any errors. But this used only for a single object.
Please advice how to fix that.

I have tried to find examples for DynamicModelMultipleChoiceField but no success.

My code.

forms.py
```from dcim.models import Device
# from netbox.forms import NetBoxModelForm
from django.forms import ModelForm, ValidationError
from utilities.forms.fields import DynamicModelChoiceField, DynamicModelMultipleChoiceField 

from .models import BarcodeList


class BarcodeForm(ModelForm):
    device = DynamicModelMultipleChoiceField(
        queryset=Device.objects.all(),
        help_text="Choose a device to add in list",
    )

    class Meta:
        model = BarcodeList
        fields = ('device',)
```
        
        
 views.py
```class DeviceBarcodeTableAddView(View):
    # class to add device to barcode bulk table form

    template_name = 'netbox_barcode/addform.html'
    form = BarcodeForm()

    def get(self, request):
        return render(
            request, self.template_name, {"form": self.form}
        )

    def post(self, request, *args, **kwargs):
        form = BarcodeForm(request.POST)
        if form.is_valid():
            form.save()
            return redirect(reverse('plugins:netbox_barcode:barcode_bulkprint'))
        return render(
            request, self.template_name, {"form": form, }
        )
```
        
        
![image](https://github.com/netbox-community/netbox/assets/94970464/0a31a51b-53c1-4b39-8d5a-e28de095f1d7)

**Comment by candlerb** [comment]
Please could you edit your post, and instead of a single backtick, put three backticks (` ``` `) on a line of their own, before and after the code. That should make it format correctly, and hence be easier to reproduce.

What version of Netbox are you using?

I have no idea why it'sgetting a ConfigContextModelQuerySet, when the queryset is Device.objects.all().  What's in `netbox_barcode/addform.html`? Could you try using your browser's developer console, and see what API calls it's making in the background?

> I have tried to find examples for DynamicModelMultipleChoiceField but no success.

There's one [here](https://docs.netbox.dev/en/stable/plugins/development/forms/#netboxmodelfiltersetform), but it's inside NetBoxModelFilterSetForm rather than NetBoxModelForm.

**Reply by candlerb** [reply]
I note that class Device inherits from ConfigContextModel, and it has this attribute:

```
    objects = ConfigContextModelQuerySet.as_manager()
```

Hence `Device.objects` is special. Indeed, nbshell confirms this:

```
>>> Device.objects
<django.db.models.manager.ManagerFromConfigContextModelQuerySet object at 0x7f560e105820>
```

But it *does* work as normal for simple queries, e.g. `Device.objects.all()`.

Hence it looks like `ConfigContextModelQuerySet` is a red herring - this is just what's normal for `Device.objects`.

The problem could just be to do with the rendering of the selection list, which should be returning integer Device IDs rather than Names for the selected values.

**Reply by aakozlov85** [reply]
Hello!
I am using 3.4.8 version of Netbox.

> What's in netbox_barcode/addform.html?

```
{% extends 'base/layout.html' %}

{% block header %}
<h1 class="barcode-title h2 w-100"> Add device to list </h1>
{% endblock header %}

{% block content %}
    <form method="POST" class="post-form">{% csrf_token %}
        {{ form.errors}}
        {% for field in form %}
            {{ field }}
            <span class='errorlist'> {{ field.errors }} </span>
            {{ field.help_text }}
        {% endfor %}
    </br>
        <button type="submit" class="btn btn-sm btn-success">Add</button>
    </form>
{% endblock %}
```

> The problem could just be to do with the rendering of the selection list, which should be returning integer Device IDs rather than Names for the selected values.

Do you mean that i should modify queryset?  Could you recommend how to do that?

**Reply by candlerb** [reply]
No, I was wondering if it might be a bug in the rendering or parsing of DynamicModelMultipleChoiceField in this context.

I can't see anything obviously wrong with what you've put, except I note that the Netbox main templates use `{% for field in form.visible_fields %}`

If you can make a minimal, complete (self-contained) reproducer, then you could raise it as an issue - especially since you say DynamicModelChoiceField works but DynamicModelMultipleChoiceField doesn't.

However, I think it would be better first to test with the current version of Netbox (3.5.6) before raising it, as there is no support for previous versions. That is: I've never seen a backport or a cut of a new version `v(X).(Y).p` after `v(X).(Y+1).0` has been released.

**Reply by aakozlov85** [reply]
I finally make it work.  but it works only when i choose one device item. not for multiple.
Now i don't know how to save form for all devices i selected.

my code:

forms.py
```
class BarcodeForm(NetBoxModelForm):
    deviceitems = DynamicModelMultipleChoiceField(
        queryset=Device.objects.all(),
        help_text="Choose a device to add in list",
    )

    class Meta:
        model = BarcodeList
        fields = ('deviceitems',)

    def save(self, *args, **kwargs):
        for item in args:
            id = item[0]
            BarcodeList.objects.create(device=id)
```

views.py
```
class DeviceBarcodeTableAddView(View):
    # class to add device to barcode bulk table form

    template_name = 'netbox_barcode/addform.html'
    form = BarcodeForm()

    def get(self, request):
        return render(
            request, self.template_name, {"form": self.form}
        )

    def post(self, request, *args, **kwargs):
        form = BarcodeForm(request.POST)
        if form.is_valid():
            deviceitems = form.cleaned_data['deviceitems']
            form.save(deviceitems)
            return redirect(reverse('plugins:netbox_barcode:barcode_bulkprint'))
        return render(
            request, self.template_name, {"form": form, }
        )
```

I found that in my POST message there are a few deviceitems (that i selected). And all with the same name.
But form adds only first device

STEP1
![image](https://github.com/netbox-community/netbox/assets/94970464/e86c22b5-56f9-449f-80b0-382fb3b32cae)

STEP2 (click ADD button)
![image](https://github.com/netbox-community/netbox/assets/94970464/f34fbc22-f8b6-43b3-9519-3c6db03121e5)

RESULT
![image](https://github.com/netbox-community/netbox/assets/94970464/51eaa6ae-a1fa-42df-badd-4b9bb4fe3e0d)

**Reply by candlerb** [reply]
"And all with the same name" - what do you see that tells you that? I see deviceItems with ids 13, 14, 15.

I think you are doing some dubious stuff here in overwriting the form's `save` method. NetBoxModelForm inherits from `django.forms.ModelForm`, and the [save method](https://docs.djangoproject.com/en/4.2/topics/forms/modelforms/#the-save-method) is only supposed to have an attribute `commit=False/True`.

You haven't shown your model `BarcodeList`, but I think you may be tying your Form to a single model instance of `BarcodeList`, when it's not.  However, this is out of my depth now so I'll stop commenting.

**Reply by aakozlov85** [reply]
Now everything works. I have been rewrite save method in my form.
My goal was to create form where i choose multiple devices from dcim.Device model, and based of device chosen, create instances for another model - BarcodeList. that s why i redefine save method.

models.py
```
class BarcodeList(models.Model):
    device = models.OneToOneField(
        to="dcim.Device",
        on_delete=models.CASCADE,
        related_name='barcode_list',
    )

    class Meta:
        verbose_name = "Barcode Device List"
        verbose_name_plural = "Barcode Device Lists"
```
